src/NHibernate/Action/CollectionRecreateAction.cs

```diff
@@ -67,15 +67,32 @@ private void PreRecreate()
 
 		public override void ExecuteAfterTransactionCompletion(bool success)
 		{
-			base.ExecuteAfterTransactionCompletion(success);
+			var cacheKey = new CacheKey(GetKey(), Persister.KeyType, Persister.Role, Session.Factory);
 			
-			if (success)
+			base.ExecuteAfterTransactionCompletion(success);
+			if (success && Persister.HasCache)
```

Comment:
`Persister.HasCache` would always be true, because otherwise the `ExecuteAfterTransactionCompletion` is not called.